services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker_background_task.py

```diff
@@ -0,0 +1,136 @@
+import asyncio
+import logging
+from datetime import datetime, timedelta, timezone
+
+from fastapi import FastAPI
+from models_library.resource_tracker import (
+    CreditTransactionStatus,
+    ServiceRunId,
+    ServiceRunStatus,
+)
+from pydantic import PositiveInt
+
+from .core.settings import ApplicationSettings
+from .models.resource_tracker_credit_transactions import (
+    CreditTransactionCreditsAndStatusUpdate,
+)
+from .models.resource_tracker_service_runs import ServiceRunStoppedAtUpdate
+from .modules.db.repositories.resource_tracker import ResourceTrackerRepository
+from .resource_tracker_utils import compute_service_run_credit_costs, make_negative
+
+_logger = logging.getLogger(__name__)
+
+_BATCH_SIZE = 20
+
+
+async def _check_service_heartbeat(
+    resource_tracker_repo: ResourceTrackerRepository,
+    base_start_timestamp: datetime,
+    resource_usage_tracker_missed_heartbeat_interval: timedelta,
+    resource_usage_tracker_missed_heartbeat_counter_fail: int,
+    service_run_id: ServiceRunId,
+    last_heartbeat_at: datetime,
+    missed_heartbeat_counter: int,
+):
+    # Check for missed heartbeats
+    if (
+        last_heartbeat_at
+        < base_start_timestamp - resource_usage_tracker_missed_heartbeat_interval
+    ):
+        missed_heartbeat_counter += 1
+        if (
+            missed_heartbeat_counter
+            > resource_usage_tracker_missed_heartbeat_counter_fail
+        ):
+            # Handle unhealthy service
+            _logger.error(
+                "Service run id: %s is considered unhealthy and not billed. Counter %s",
+                service_run_id,
+                missed_heartbeat_counter,
+            )
+            await _close_unhealthy_service(
+                resource_tracker_repo, service_run_id, base_start_timestamp
+            )
+        else:
+            _logger.warning(
+                "Service run id: %s missed heartbeat. Counter %s",
+                service_run_id,
+                missed_heartbeat_counter,
+            )
+            await resource_tracker_repo.update_service_missed_heartbeat_counter(
+                service_run_id, last_heartbeat_at, missed_heartbeat_counter
+            )
+
+
+async def _close_unhealthy_service(
+    resource_tracker_repo, service_run_id, base_start_timestamp
+):
+    # 1. Close the service_run
+    update_service_run_stopped_at = ServiceRunStoppedAtUpdate(
+        service_run_id=service_run_id,
+        stopped_at=base_start_timestamp,
+        service_run_status=ServiceRunStatus.ERROR,
+        service_run_status_msg="Service missed more heartbeats. It's considered unhealthy.",
+    )
+    running_service = await resource_tracker_repo.update_service_run_stopped_at(
+        update_service_run_stopped_at
+    )
+
+    if running_service is None:
+        _logger.error(
+            "Service run id: %s Nothing to update. This should not happen; investigate.",
```

Comment:
I guess that is one of these errors, that should be catched by graylog. Check with the devops to get that in place...

Reply:
Ok, i will check with them ... btw my idea is that any error from RUT or Payment Service should go to mattermost specific billing error channel